src/app/reporting/Engine.cpp

```diff
@@ -336,6 +367,65 @@ void Engine::Run()
         mCurReadHandlerIdx = (mCurReadHandlerIdx + 1) % CHIP_IM_MAX_NUM_READ_HANDLER;
         readHandler        = imEngine->mReadHandlers + mCurReadHandlerIdx;
     }
+    for (auto & path : mChangeSetPathList)
+    {
+        path.ClearDirty();
+    }
+}
+
+void Engine::SetDirty(ClusterInfo & aClusterInfo)
+{
+    for (auto & path : mChangeSetPathList)
+    {
+        if (path.IsDirty())
+        {
+            // If aClusterInfo is part of existing dirty path, skip it
+            // If aClusterInfo is super set of one existing path, update that path with new path.
+            if (path.IsAttributePathSupersetOf(aClusterInfo))
+            {
+                return;
+            }
+            if (aClusterInfo.IsAttributePathSupersetOf(path))
+            {
+                path.mListIndex = aClusterInfo.mListIndex;
+                path.mFieldId   = aClusterInfo.mFieldId;
+                path.mFlags     = aClusterInfo.mFlags;
+                return;
+            }
+        }
+    }
+
+    for (auto & readHandler : InteractionModelEngine::GetInstance()->mReadHandlers)
+    {
+        if (readHandler.IsSubscription() && (readHandler.IsReportable() || readHandler.IsReporting()))
+        {
+            ClusterInfo * clusterInstance = readHandler.GetAttributeClusterInfolist();
+            while (clusterInstance != nullptr)
+            {
+                // If dirty path intersects with interested path from handler, it mark this in ChangeSetPathList and also
+                // mark cluster as dirty for the particular subscribe handler , otherwise, it skip this dirty path.
+                if (clusterInstance->IsAttributePathSupersetOf(aClusterInfo) ||
+                    aClusterInfo.IsAttributePathSupersetOf(*clusterInstance))
+                {
+                    for (auto & path : mChangeSetPathList)
```

Comment:
Except for this for loop, I don't think the giant outer for loop here starting from 398 to 425 is needed - it's pre-mature optimization to need to mark the paths stored in the ReadHandlers as dirty _in addition to_ the core dirty set tracked in the engine. In your implementation in `BuildSingleReportDataAttributeDataList`, you already do this permuted-path-intersection check, so doing it extra here doesn't buy you much. I'd rather this logic be simpler to read _for starters_, and then optimize if we think we need to later.

The other nice side-effect of removing this optimization is that you can then delete the dirty flag from within `ClusterInfo`, which then pivots it to be purely a data structure that stores a path.

Reply:
agree, will do it, thanks